### PR TITLE
feat(data-pipeline): copy_dataset_root_uri accepts file:// and r2://

### LINKED ISSUE #1547: feat(data-pipeline): copy_dataset_root_uri accepts file:// and r2:// copy sources
## Summary

Widen the dataset-copy source from a worker-local directory to a **root URI**, mirroring the finalize-side change in #1542 (run-prefix URIs end to end).

Rename `DatasetSpec.copy_dataset_root` → `copy_dataset_root_uri` and accept a bare path, `file://` URI, or `r2://` URI. An `r2://` source shard is downloaded to a tempfile per shard at copy time, so the source no longer has to be synced locally before a copy run.

## Motivation

Copy runs currently require the operator to `rclone sync` the source dataset locally first. Accepting an `r2://` root lets `generate_dataset` pull source shards on demand — the same convention the datamodule (`download_dataset_root_uri`) and finalize (`dataset_root_uri`, #1542) already use.

## Scope

- `copy_dataset_root` → `copy_dataset_root_uri` on `DatasetSpec`, `dataset.yaml`, and the `--copy_dataset_root_uri` CLI forward.
- Per-shard source resolved through `localized_uri` (downloads `r2://` shards to a tempfile).
- Launch-time `_validate_copy_source` loads the source `input_spec.json` via `load_spec_from_uri`, so an `r2://` source spec validates without a local sync.
- Back-compat shim promotes both pre-rename shapes (flat `copy_dataset_root`, nested `datasetsrc: {copy_dataset_root: …}`); `from_hydra_cfg` rejects either stale Hydra key with a migration pointer.
- Docs (`data-pipeline.md`) updated.

### PR DESCRIPTION
**Pull request by ktinubu**
## Summary

Widens the dataset-copy source from a worker-local directory to a **root URI**, the generate-side sibling of the finalize change in #1542 (run-prefix URIs end to end).

- Renames `DatasetSpec.copy_dataset_root` → `copy_dataset_root_uri` across the schema, `dataset.yaml`, and the `--copy_dataset_root_uri` CLI forward.
- The field now accepts a bare path, `file://` URI, or `r2://` URI. An `r2://` source shard is resolved through `localized_uri` and downloaded to a tempfile per shard at copy time, so the source no longer has to be `rclone sync`'d locally before a copy run.
- Launch-time `_validate_copy_source` loads the source `input_spec.json` via `load_spec_from_uri`, so an `r2://` source spec validates at launch without a local sync.

## Back-compat

- A `mode="before"` validator promotes both pre-rename shapes to `copy_dataset_root_uri` — the flat `copy_dataset_root` and the older nested `datasetsrc: {copy_dataset_root: …}` (and drops `datasetsrc: null`).
- `from_hydra_cfg` rejects either stale Hydra key with a migration pointer.

## Testing

- `make test-fast` green.
- New behavioral coverage: `localized_uri` scheme dispatch (bare / `file://` / `r2://` / unsupported), the r2:// per-shard download path end to end, legacy-shape promotion for both shapes, blank-rejection, and the hdf5-only guard.

## Review

Pre-PR `/repo-review-full-no-comments`: 0 BLOCK, comment-hygiene clean. 11 WARN advisories (error-message specificity in `_validate_copy_source`, a shared URI-join helper, and a few coverage gaps) are deferred as inline review follow-ups.

Fixes #1547
Refs #1542

### CODE REVIEW

**Comment by ktinubu** [comment]
## Verification Results

Verified at HEAD `bb40262b04ab028620bd1a168cec4cd92c3253f8` (matches PR head) in the devcontainer worktree, with Surge XT VST3 linked (`plugins/Surge XT.vst3` → `/usr/lib/vst3/Surge XT.vst3`).

- [x] **Copying a dataset reproduces the source's params end-to-end** *(Level 1 — promise is end-to-end; renders a real source shard via Surge XT, decodes its `param_array`, re-renders from the fixed params, and asserts equality)*

  ```bash
  $ src/synth_setter/scripts/run-linux-vst-headless.sh .venv/bin/pytest -vv -s \
      tests/data/vst/test_generate_vst_dataset.py::test_copy_dataset_reproduces_source_param_array
  ...
  40.32s call  tests/data/vst/test_generate_vst_dataset.py::test_copy_dataset_reproduces_source_param_array
  ============================== 1 passed in 41.41s ==============================
  ```

- [x] **`copy_dataset_root_uri` resolves a bare path / `file://` in place and rejects unsupported schemes** *(Level 2 — invokes `localized_uri` directly)*

  ```bash
  $ .venv/bin/python3 -c "
  from pathlib import Path
  from synth_setter.pipeline.spec_io import localized_uri
  p = Path('/tmp/loc_uri_probe.txt'); p.write_text('hi')
  with localized_uri(str(p)) as g:        print('bare   ->', g, 'in place:', g == p)
  with localized_uri(p.as_uri()) as g:    print('file://->', g, 'decoded :', g == p)
  for bad in ('s3://bucket/x','http://h/x'):
      try:
          with localized_uri(bad): pass
      except ValueError as e: print(bad, '->', str(e)[:48])
  "
  bare   -> /tmp/loc_uri_probe.txt in place: True
  file://-> /tmp/loc_uri_probe.txt decoded : True
  s3://bucket/x -> unsupported URI scheme 's3': 's3://bucket/x'.
  http://h/x -> unsupported URI scheme 'http': 'http://h/x'.
  ```

- [x] **An `r2://` source shard is downloaded to a tempfile per shard (and cleaned up), and the CLI forwards it through render** *(Level 2 — `TestLocalizedUri` r2 path drives the download+cleanup; the CLI `main()` test stubs rclone to serve a fixture and asserts decoded params reach the writer)*

  ```bash
  $ .venv/bin/pytest -q \
      tests/pipeline/test_spec_io.py::TestLocalizedUri \
      tests/data/vst/test_generate_vst_dataset.py::test_main_copy_dataset_root_uri_downloads_r2_source_shard
  # includes: test_r2_uri_downloads_to_tempfile_and_cleans_up, test_unsupported_scheme_is_rejected_with_value_error,
  #           test_bare_local_path_yields_path_in_place, test_file_uri_yields_decoded_local_path,
  #           test_main_copy_dataset_root_uri_downloads_r2_source_shard  → all passed
  ```

- [x] **Legacy `copy_dataset_root` and nested `datasetsrc` shapes promote to `copy_dataset_root_uri`; ambiguous/blank/`wds` inputs are rejected** *(Level 2 — `TestCopyDatasetRootUri` runs the real `mode="before"` validator over in-memory and `model_validate_json` inputs)*

  ```bash
  $ .venv/bin/pytest -q tests/pipeline/schemas/test_dataset_spec.py::TestCopyDatasetRootUri
  # incl. test_legacy_flat_copy_dataset_root_is_promoted, test_legacy_datasetsrc_mapping_is_promoted,
  #       test_legacy_datasetsrc_null_is_dropped, test_both_legacy_datasetsrc_and_uri_keys_are_rejected,
  #       test_copy_dataset_root_uri_blank_is_rejected, test_copy_dataset_root_uri_with_wds_output_is_rejected,
  #       *_json_loads_via_model_validate_json (both shapes)  → all passed (22 passed in 1.09s for the full selection)
  ```

- [x] **`from_hydra_cfg` rejects both stale keys (`datasetsrc`, `copy_dataset_root`) with a migration pointer** *(Level 2 — `TestFromHydraCfg` composes an OmegaConf cfg carrying the stale key and asserts the `ValueError`)*

  ```bash
  $ .venv/bin/pytest -q tests/pipeline/schemas/test_dataset_spec.py -k "hydra and (stale or datasetsrc or copy_dataset_root)"
  TestFromHydraCfg::test_stale_copy_dataset_root_key_is_rejected_with_migration_pointer PASSED
  TestFromHydraCfg::test_stale_datasetsrc_key_is_rejected_with_migration_pointer PASSED
  2 passed, 121 deselected in 0.45s
  ```

### Not verified (adversarial gap — matches pre-PR review WARN)

- [ ] **An `r2://` copy root with a missing/unreadable remote `input_spec.json` yields the friendly "sync the source spec" `ValueError` (not a raw rclone error)** *(Level 2 — SKIP: no test exercises the `subprocess.CalledProcessError` arm of `_validate_copy_source`; only the local `FileNotFoundError` arm is covered. A mutant dropping `subprocess.CalledProcessError` from the except tuple would survive.)*
  To cover this, add a CLI-preflight test pointing `copy_dataset_root_uri` at an `r2://` root and stub the rclone download to exit non-zero:
  ```bash
  # in tests/pipeline/entrypoints/test_generate_dataset_unit.py::TestValidateCopySource
  # monkeypatch r2_io's check_call to raise subprocess.CalledProcessError, then assert
  # _validate_copy_source(spec) raises ValueError matching INPUT_SPEC_FILENAME.
  ```
  Note: a malformed source spec still surfaces as a raw `pydantic.ValidationError` here (pre-existing behavior, not a regression).

**Review by ktinubu — COMMENTED** [review]
⛔ 3 BLOCKING finding(s) — changes required (self-review: posted as COMMENT)

Multi-skill review of PR #1548 — 7 parallel passes (code-health, synth-setter-project-standards, python-style, tdd-implementation, comment-hygiene, ml-data-pipeline, ml-test). No diff-anchored BLOCK findings were raised by any skill, so there are no inline threads; every advisory WARN is collapsed into the `## Advisory (WARN) findings` section below to keep the list short. The only blocking items are PR-health failing checks under `## PR health` (independent of the diff).

## PR health

- **[repo-review-full:block]** [pr-health] Failing check: Generate + finalize smoke (hdf5) / Run generate_dataset (skypilot-local) (FAILURE) — https://github.com/tinaudio/synth-setter/actions/runs/27075746398/job/79912737172
- **[repo-review-full:block]** [pr-health] Failing check: lock-check (ubuntu-latest) (FAILURE) — https://github.com/tinaudio/synth-setter/actions/runs/27075746340/job/79912736758
- **[repo-review-full:block]** [pr-health] Failing check: lock-check (macos-latest) (FAILURE) — https://github.com/tinaudio/synth-setter/actions/runs/27075746340/job/79912736755

## Advisory (WARN) findings

- **[code-health:warn]** src/synth_setter/cli/generate_dataset.py:67 — DRY: `_validate_copy_source` re-implements the root->`input_spec.json` join+load already encapsulated in `spec_io.load_spec_from_root`; reuse it inside the try/except to drop the duplicated knowledge of where the spec sits relative to the root.
- **[code-health:warn]** src/synth_setter/cli/generate_dataset.py:70 — The catch maps `subprocess.CalledProcessError` to "source has no input_spec.json", but rclone exits non-zero on auth/network failures too — conflates a missing spec with an environmental failure and misdirects the operator (the real cause is still chained via `from exc`) (also: ml-pipeline).
- **[code-health:warn]** tests/pipeline/entrypoints/test_generate_dataset_unit.py:2801 — `test_missing_source_spec_raises_with_path` name overstates the contract: the new message interpolates the root URI, not the resolved `source_spec_path`; rename to `_raises_with_root_uri` and assert on the URI.
- **[synth-setter:warn]** src/synth_setter/data/vst/generate_vst_dataset.py:199 — Per-shard `r2://` download (via `localized_uri`->`downloaded_to_tempfile`->`download_to_path`) runs `rclone copyto --checksum` with no `--retries`/timeout, unlike sibling r2 helpers; now on the renderer hot path, so a single transient blip fails that shard's render with an unhandled traceback. Integrity is fine (`--checksum`); resilience and an actionable message are the gap (also: ml-pipeline).
- **[tdd-impl:warn]** src/synth_setter/cli/generate_dataset.py:67 — No test in `TestValidateCopySource` exercises an `r2://` copy root; every case reaches the `FileNotFoundError` arm, so the `CalledProcessError -> ValueError` translation (the integration seam this PR adds) is untested. Add an r2 case that stubs the rclone download to fail, plus an r2 happy path (also: ml-test).
- **[tdd-impl:warn]** tests/data/vst/test_generate_vst_dataset.py:563 — The r2 tests patch `r2_io.subprocess.check_call` with a `shutil.copy`/`write_text` stub instead of the higher-fidelity `fake_r2_remote` fixture (tests/pipeline/conftest.py:15), bypassing real rclone argv/flag construction; the stub also couples to `args[-1]` positional ordering (also: ml-test).
- **[tdd-impl:warn]** tests/pipeline/test_spec_io.py:999 — `test_r2_uri_downloads_to_tempfile_and_cleans_up` asserts cleanup only on the success path; the failure-path contract (an exception inside the `with` body still removes the tempfile) is untested.
- **[ml-pipeline:warn]** src/synth_setter/pipeline/spec_io.py:445 — `localized_uri` yields bare local/`file://` paths with no existence check, so a missing copy source surfaces later at the `h5py`/`read_text` call and loses the "this is the copy source that's missing" context the r2 branch carries. Minor.
- **[ml-test:warn]** src/synth_setter/data/vst/generate_vst_dataset.py:198 — The failing r2-fetch path in `main()` (uncaught `CalledProcessError` mid-render) has no covering test; only the happy-path download is tested.
- **[ml-test:warn]** tests/pipeline/test_spec_io.py:1013 — The malformed-`file://` `ValueError` advertised in `localized_uri`'s docstring (propagated from `file_uri_to_path`) is untested; only the unsupported-scheme `ValueError` is exercised.

File: src/synth_setter/cli/generate_dataset.py [hunk 1]
```diff
@@ -251,37 +251,40 @@ def build_generate_args(spec: DatasetSpec, shard: ShardSpec, output_dir: Path) -
...
-            f"dataset-copy source has no {INPUT_SPEC_FILENAME} at {source_spec_path}; "
-            "sync the source dataset's spec alongside its shards (it lives beside the "
-            "shards at the R2 dataset prefix root) so the copy can be validated."
-        )
-    source = load_spec_from_uri(str(source_spec_path))
+            f"dataset-copy source has no {INPUT_SPEC_FILENAME} under "
+            f"{spec.copy_dataset_root_uri!r}; sync the source dataset's spec alongside its "
+            "shards (it lives beside the shards at the dataset prefix root) so the copy "
+            "can be validated."
+        ) from exc
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply src/synth_setter/cli/generate_dataset.py:293]: Fixed in 6f49f70e. Split the `except` in `_validate_copy_source`: a genuinely-missing spec (`FileNotFoundError`) keeps the "sync the source spec" message, while a non-zero rclone exit (`subprocess.CalledProcessError` from the `r2://` fetch) now raises a distinct `ValueError` that names the failing rclone command and exit code and points at auth/network/config — no longer masked as "missing input_spec.json". New test `test_rclone_fetch_failure_raises_distinct_from_missing_spec` stubs rclone to exit non-zero and asserts the distinct message plus the chained `CalledProcessError` cause.

File: src/synth_setter/pipeline/spec_io.py [hunk 2]
```diff
@@ -56,33 +59,51 @@
...
+
+    :param uri: Local filesystem path, ``file://`` URI, or ``r2://`` URI.
+    :yields Path: A local path readable for the duration of the context.
+    :raises ValueError: ``uri`` carries a scheme other than ``file://`` or
+        ``r2://``, or is a malformed ``file://`` URI (propagated from
+        :func:`~synth_setter.pipeline.file_uri.file_uri_to_path`).
+    """
+    scheme = urlparse(uri).scheme
+    if scheme in _LOCAL_FILESYSTEM_SCHEMES:
+        yield file_uri_to_path(uri) if scheme == "file" else Path(uri)
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply src/synth_setter/pipeline/spec_io.py:99]: Left as-is. Confirmed the parse — `urlparse("C:\\data\\shard.h5").scheme == "c"`, so this would indeed raise `ValueError` on a Windows drive-letter path.

But synth-setter is a Linux-only pipeline (SkyPilot → RunPod/OCI compute, Cloudflare R2 storage; no Windows runtime or CI), and `localized_uri`'s bare-path semantics follow the POSIX rclone/fsspec/Arrow convention used everywhere here. Drive-letter inputs never reach this code path, so adding Windows normalization would be speculative handling for an unsupported platform (YAGNI). Happy to revisit if a Windows target is ever added.
